database/storage_engine.c

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "storage_engine.h"
+#include "rrddim_mem.h"
+#ifdef ENABLE_DBENGINE
+#include "engine/rrdengineapi.h"
+#endif
+
+static void dimension_destroy_freez(RRDDIM *rd) {
+    freez(rd);
+}
+static void dimension_destroy_unmap(RRDDIM *rd) {
+    munmap(rd, rd->memsize);
+}
+static void set_destroy_freez(RRDSET *st) {
+    freez(st);
+}
+static void set_destroy_unmap(RRDSET *st) {
+    munmap(st, st->memsize);
+}
+
+// Initialize a rrdset with calloc and the provided memory mode
+static RRDSET* set_init_calloc(RRD_MEMORY_MODE mode) {
+    RRDSET* st = callocz(1, sizeof(RRDSET));
+    st->rrd_memory_mode = mode;
+    return st;
+}
+
+#define SET_INIT(MMODE) \
+static RRDSET* rrdset_init_##MMODE(const char *id, const char *fullid, const char *filename, long entries, int update_every) { \
+    return set_init_calloc(RRD_MEMORY_MODE_##MMODE); \
+}
+SET_INIT(NONE)
+SET_INIT(ALLOC)
+SET_INIT(DBENGINE)
+
+// Initialize a rrddim with calloc and the provided memory mode
+static RRDDIM* dim_init_calloc(RRDSET *st, RRD_MEMORY_MODE mode) {
+    size_t size = sizeof(RRDDIM) + (st->entries * sizeof(storage_number));
+    RRDDIM* rd = callocz(1, size);
+    rd->memsize = size;
+    rd->rrd_memory_mode = mode;
+    return rd;
+}
+
+#define DIM_INIT(MMODE) \
+static RRDDIM* rrddim_init_##MMODE(RRDSET *st, const char *id, const char *filename, collected_number multiplier, collected_number divisor, RRD_ALGORITHM algorithm) { \
+    return dim_init_calloc((st), RRD_MEMORY_MODE_##MMODE); \
+}
+DIM_INIT(NONE)
+DIM_INIT(ALLOC)
+DIM_INIT(DBENGINE)
+
+static const struct rrddim_collect_ops im_collect_ops = {
+    .init = rrddim_collect_init,
+    .store_metric = rrddim_collect_store_metric,
+    .finalize = rrddim_collect_finalize
```

Comment:
Functions that start with `rrddim_<verb>_...` imply that they are storage-engine independent, whereas in this case these refer to the memory implementation. Would it be possible to use a dedicated prefix for each separate memory mode (or one for RAM-modes/dbengine/mongodb/etc.)?

The intention of my comment is to end up with something like:

`.init = storage_engine_ram_rrdim_init`,
`.init = storage_engine_dbeng_rrddim_init`,
`.init = storage_engine_mongodb_rrddim_init`.

Beware, this is just a quick suggestion from my part. Considering that naming stuff is difficult, feel free to use any scheme that you might prefer.

Reply:
I agree with the suggestion. I'll rename these functions while introducing the new dim/set APIs in part 3.